### DISCUSSION TITLE
UM FeatherS3 - has anyone tested these? Or tested ESPNow with ESP32-S3?

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#12017] [ESP32]
I've been working on ESPNow for a while with erratic results from these boards. They don't seem to like communicating, going through periods of repeated ESPNow transmission failures. Replacing the Feather with the ESP32 reference board the problem goes away with thousands of messages being passed with very rare errors.

The fault is highly variable. At one point I convinced myself I had one bad board, when another Feather worked fine for a few days. Then it started dropping messages. The Feather is a nice board - aside from this I can't fault the hardware; but if it's not capable of reliable communication it's a turkey.

I'm unsure if the issue is specific to the Feather, or if it's an S3 problem, or whether it's associated with SPIRAM. I'm currently testing an ESP32 with SPIRAM which is working perfectly so far.

I'd be glad of any feedback on this.

**Comment by beetlegigg** [comment]
I got a Feather S3 from UM shortly after they were first produced.  I had an issue with reliable wifi connection that showed up when I was testing that my wifi code on the board was reliably working when testing that the board would powering up and run after a power cut.  It turned out that several of these boards were having wifi issues and it was replace with a new and tested by UM board via pimoroni where I purchased the board.    The replacement board seems OK, but I moved to project to a rpi pico and have not used the Feather S3 much since then.   When I had the wifi issue I reached out to UM on their discord site, and Seon was very helpful.   It would probably be worth your while connecting to Seon directly to see what his thoughts are.

**Reply by peterhinch** [reply]
Thanks for that.

**Comment by UnexpectedMaker** [comment]
Hey @peterhinch - Sorry to hear you are potentially having some issues with your FS3.

Firstly, can you let me know what the `P` number is on the back of the board under the USB connector?  It should be P4, P7 or P8.

Also, when you say it work's ok with an ESP32 reference board, do you mean an ESP32-S3 DevKit board? or just an original ESP32 board?

The ESP32-S3 it quite a temperamental beast compared to the original ESP32 and even the ESP32-S2.  It has radio issues when native USB is active and you are using wifi channels below Channel 9 (degraded TX strength due to native USB interference) - and it's also got secondary harmonics issues and can switch TX off when you try to use max strength on the radios.

That said, you should not be seeing stability issues  - It should either work or not work in a situation - at least from my experience. 

Cheers,
Seon

**Reply by peterhinch** [reply]
I have three FS3 boards, two are P7 and one is P4.

The reference board is an original ESP32-Devkit-C. I have also run other original ESP32 boards - all work faultlessly.

From your description the ESP32-S3 device sounds like a turkey. This is a shame because I really like the design of the FS3 board, with the switchable regulator and the board's phenomenally low deepsleep current.

To describe my test setup, I have an original ESP32 acting as an MQTT gateway. It provides an MQTT subset to multiple nodes which communicate with it using ESPNow. In development I run a single node attached to my laptop by USB, which is in the next room from the gateway. The problem I get is with `ESPNow.send` failing. With an original ESP32 I have sent 18,000 messages without a failure. On another run I might see one failure in 1,000 messages. With the FS3, some days it will work. Other days I will see almost as many failures as successes.

I am using channel 3. I have had problems with all three FS3 boards (P4 and P7).

**Comment by UnexpectedMaker** [comment]
Thanks for the clarification, Yeah, P4 ad P7 are the same on the RF side. The changes were only DFM things and moving the second STEMMA/QT onto different IO to allow different I2C busses on each STEMMA/QT connector.

Ok, first thing you should try is using channel 9 or above - you should see an improvement just from that if what you are seeing is USB interference related.  For reference, in MP and CP, USB is enabled all of the time, even if you are not plugged in via USB or have a REPL open, so it will always be agitating the lower channels - by how much is unknown - I've been told 1-2dB loss from some Espressif folks and 3-4dB from others :(

The other thing you might want to try (might not be possible from within MP as I've not looked at the new ESPNow stuff) is to reduce the TX output strength down to a lower value - in the IDF, 20 is the max (and I believe the default) but setting it to 16 or 17 might help with the stability.

If you want to send me a private message with your address, I can send you a new P8 revision board to try out - it has a new antenna & RF layout due to the white Ignion antennas I was using going EOL with out a drop in replacement.

I'd be curious to see if the P8 revision works any better in your setup - I have broadened the freq bandwidth on the P8 version to help compensate for the USB issue, but of course that also means a slight reduction in dB overall, but that should be unnoticeable for general WiFi related projects.

Cheers :)

**Reply by glenn20** [reply]
> The other thing you might want to try (might not be possible from within MP as I've not looked at the new ESPNow stuff) is to reduce the TX output strength down to a lower value - in the IDF, 20 is the max (and I believe the default) but setting it to 16 or 17 might help with the stability.

Yes, espnow will just use any `txpower` setting set in the wifi setup, so `network.WLAN(network.STA_IF).config(txpower=17)` will set the maximum txpower. (Note due to the wierd mappings used, I believe setting `txpower=17` will actually yield a value of `16` due to the [actual txpower mappings here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_wifi.html#_CPPv425esp_wifi_set_max_tx_power6int8_t) (**if the docs are correct and I've followed them correctly**). Note that micropython mulltiplies the supplied value by 4 and passes that to `esp_wifi_set_max_tx_power()`. You can check, by calling `network.WLAN(network.STA_IF).config("txpower")` after setting the value.

**Comment by peterhinch** [comment]
@UnexpectedMaker Thanks for the suggestions. I will try channel 9 and `network.WLAN(network.STA_IF).config(txpower=17)`. 

It may be a week or so before I can be confident of a positive outcome because the appearance of the fault is so erratic.

I have sent you an email with my address. Thank you for your help.

**Comment by tferrin** [comment]
@peterhinch, Did setting a higher channel number and reducing the TX power level improve the reliability in your ESPNow experiments? I have a number of UM FeatherS3 boards and would love to use them in an ESPNow deployment.

**Reply by peterhinch** [reply]
Check the version number on the underside next to the USB connector. P8 or R8 boards work on any channel without special precautions. P4 and P7 boards work fine on a high channel with reduced power.

**Comment by BKneer** [comment]
Hello, not sure if it helps... or if someone has made similar experience. I was experimenting with esp-now basic setup. The communication from a ESP32-s3 zero or ESP32-c3 super mini to another ESP32-c3 super mini (this is the generel receiver) works fine but with two other ESP32-s3 zero I see a continuous reset loop. After some investigation I've figured out that these ESPs loose their esp-now config completely. If I as in the loop for the peer number I always get 0. Even if I set the peer again I see resets if I don't detach and init the esp-now connection again... 
If I do the init, callback register, and peer setting in the loop the sending works stable.
This is really strange and I would like to understand why these esp32-s3 zeros loose the esp-now configuration (mac only differ in the 6th byte) from one loop to the next. As mentioned above it works for another ESP32-s3 zero (different mac bytes 4-6) , ESP32-c3 and an ESP32 DEV...
